sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/GpuWindowGroupLimitExec.scala

```diff
@@ -0,0 +1,263 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "350"}
+{"spark": "351"}
+spark-rapids-shim-json-lines ***/
+package com.nvidia.spark.rapids.shims
+
+import ai.rapids.cudf
+import ai.rapids.cudf.GroupByOptions
+import com.nvidia.spark.rapids.{BaseExprMeta, DataFromReplacementRule, GpuBindReferences, GpuBoundReference, GpuColumnVector, GpuExec, GpuExpression, GpuMetric, GpuOverrides, GpuProjectExec, RapidsConf, RapidsMeta, SparkPlanMeta, SpillableColumnarBatch, SpillPriorities}
+import com.nvidia.spark.rapids.Arm.withResource
+import com.nvidia.spark.rapids.RmmRapidsRetryIterator.{splitSpillableInHalfByRows, withRetry}
+import com.nvidia.spark.rapids.window.{GpuDenseRank, GpuRank}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, DenseRank, Expression, Rank, SortOrder}
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.window.{WindowGroupLimitExec, WindowGroupLimitMode}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+sealed trait RankFunctionType
+case object RankFunction extends RankFunctionType
+case object DenseRankFunction extends RankFunctionType
+
+class GpuWindowGroupLimitExecMeta(limitExec: WindowGroupLimitExec,
+                                  conf: RapidsConf,
+                                  parent:Option[RapidsMeta[_, _, _]],
+                                  rule: DataFromReplacementRule)
+    extends SparkPlanMeta[WindowGroupLimitExec](limitExec, conf, parent, rule) {
+
+  private val partitionSpec: Seq[BaseExprMeta[Expression]] =
+    limitExec.partitionSpec.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+
+  private val orderSpec: Seq[BaseExprMeta[SortOrder]] =
+    limitExec.orderSpec.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
+
+  private val rankFunction = GpuOverrides.wrapExpr(limitExec.rankLikeFunction, conf, Some(this))
+
+  override def tagPlanForGpu(): Unit = {
+    wrapped.rankLikeFunction match {
+      case DenseRank(_) =>
+      case Rank(_) =>
+      // case RowNumber() => // TODO: Future.
+      case _ => willNotWorkOnGpu("Only Rank() and DenseRank() are " +
+                                 "currently supported for window group limits")
+    }
+  }
+
+
+  override def convertToGpu(): GpuExec = {
+    GpuWindowGroupLimitExec(partitionSpec.map(_.convertToGpu()),
```

Comment:
I am a little confused by this. The code looks okay, but why are we not reusing all of the code that we have for GpuRunningWindowExec?  The only reason this works is because the operators are running window operators that are computed based on position more than anything else.  We know that the existing code works for that, so it really would end up being GpuRunningWindowExec(but no input partitioning requirement) -> Filter -> Project (to drop the rankLikeFunction column)

I am fine if we hide all of that in a single `GpuWindowGroupLimitExec`, but I really would like to see the internals look something more like.

```
child.executeColumnar().mapPartitions { iter =>
      val running = new GpuRunningWindowIterator(iter, boundWindowOps, boundPartitionSpec, boundOrderSpec,
        output.map(_.dataType).toArray, numOutputBatches, numOutputRows, opTime)
      running.map(filterAndDropRank)
    }
```

This would let us deal with range and row based frames while not needing to keep an entire window in memory.  We could even add in more optimizations later that would pre-filter input batches with the same partition by parameters as batches that we already satisfied.

Reply:
> why are we not reusing all of the code that we have for GpuRunningWindowExec?
> ... This would let us deal with range and row based frames 

I started out thinking I'd reuse the innards of `GpuRunningWindowExec`. There were a couple of reasons against:
1. All the ranking functions have a fixed window definition: `[UNBOUNDED PRECEDING, CURRENT ROW]`.  They cannot be `RANGE` frames either; they're always `ROW` based.
2. The running window optimization is not necessary for the filter optimization to work. If the window spans multiple batches, we can still filter all but `N` rows from each of the batches.  When the actual window finally runs, the final `N` will be filtered in.  Carrying state between batch computations isn't required.
3. Only the basic window computation would be critical to use from `GpuRunningWindowExec`.

Given this, I figured it would be easiest to compute the rank via scans.

Reply:
To be clear I am mostly trying to understand your thought process for this. My main concern with this is just that it is different from how Spark does it. For good or bad. But that is very much a nit. The current code adds only 150 extra lines and is very simple so I am fine with it.

Reply:
Thanks, @revans2.  I should document the reasoning a little bit here.

The Apache Spark approach seems to use a [row-wise iterator](https://github.com/apache/spark/blob/f0090c95ad4eca18040104848117a7da648ffa3c/sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala#L114) that keeps track of the current row rank (based on the group/order-by column values).

By my read, we can't take the same approach in a columnar system like ours; we'd have to process the whole column.  And to also compare grouping/ordering column values, using the rank scans made sense.

I'll document this reasoning in the class.

Reply:
Added documentation clarifying the approach.